opte/src/engine/packet.rs

```diff
@@ -2622,4 +2620,48 @@ mod test {
         assert_eq!(pkt.num_segs(), 2);
         assert!(matches!(pkt.parse(), Err(ParseError::BadHeader(_))));
     }
+
+    // Verify that we correctly parse an IPv6 packet with extension headers
+    #[test]
+    fn parse_ipv6_extension_headers_ok() {
+        use crate::engine::ip6::test::generate_test_packet;
+        use crate::engine::ip6::test::SUPPORTED_EXTENSIONS;
+        use itertools::Itertools;
+        use smoltcp::wire::IpProtocol;
+        for n_extensions in 0..SUPPORTED_EXTENSIONS.len() {
+            for extensions in
+                SUPPORTED_EXTENSIONS.into_iter().permutations(n_extensions)
```

Comment:
This is certainly a nice way to automate testing various permutations, my only question is if the IPv6 spec allows arbitrary combinations? Like, is there any reason we should fail to parse a packet given a certain combination/ordering of extension headers?

Reply:
Good question. On my first reading of the [relevant section](https://www.rfc-editor.org/rfc/rfc2460#section-4.1) of the RFC, I focused on this:

> When more than one extension header is used in the same packet, it is
   recommended that those headers appear in the following order:

I took "recommended" to imply that this is just a recommendation, not a requirement for a conforming implementation. On closer reading, I think you're right, that there is a required order. Specifically, Hop-By-Hop headers are required to appear just after the base header (if they're present at all), and only once:

> IPv6 nodes must accept and attempt to process extension headers in
   any order and occurring any number of times in the same packet,
   except for the Hop-by-Hop Options header which is restricted to
   appear immediately after an IPv6 header only.

In that case, what's the right behavior should we receive a packet that doesn't conform? I imagine it's to drop it, but I saw elsewhere in the IPv6 spec that sometimes sending a special ICMP packet back to the sender is required. That's mentioned as a valid response to a Next Header that the receiver doesn't support, so I'm not sure it applies here.

Reply:
Okay, so I read the RFC a bit and now I'm thinking we actually shouldn't care about order, and we should pass on packet with the extension headers as-is. If they are bogus, then either some router along the path or the destination itself will drop the packet and perhaps send an "ICMP Parameter Problem" message back in reply. OPTE is not really in the business of enforcing strictness of the headers. Yes, it does do some validation to make sure the headers are sane, but verifying order of extension headers feels like something else, at least for the guest packets.

Reply:
We've agreed offline that we definitely don't want OPTE to be in the business of verifying EH order right now, and probably never. As @rzezeski notes, if a client of OPTE sends us a packet that has out-of-order headers, they're kinda on their own. Either they're IP stack is broken, or they're doing something funky / custom and it's up to them to make sure they'll get whatever replies they need, i.e., that the broken packets will make it to their destination and back.